### ISSUE TITLE
Allow running a Function for more than 5 minutes on a Consumption plan

### ISSUE THREAD

**Issue by lindydonna** [Azure/Azure-Functions#75]
Copied from: https://github.com/Azure/azure-webjobs-sdk-script/issues/18

The underlying compute platform for Consumption plan Functions does not allow running for more than 5 minutes, so the Functions host enforces this limit. There's customer feedback that it would be useful to increase this timeout.

This issue tracks work that will have to be done across the compute platform, the Functions host, and (possibly) the portal.

**Comment by mamaso** [comment]
Leaving this relevant customer post here: http://stackoverflow.com/questions/40800933/azure-functions-pricing-and-timeout

**Comment by mikezks** [comment]
I agree with the initial SO post: Serverless means for me, that I do not have to take care about instances, sizing up resp. down and have a clear pricing model to calculate the costs for running a job - no matter if it needs 10 sec or 5 days.

It would be good to have the choice of setting a timeout, be informed about the costs per email periodically after each definable interval, etc.

But using ARM to activate and deactivate a Function App within a Web App does not follow the principles of a serverless environment. I use these kind of services because I am convinced that Microsoft can do a better job concerning operation than I do myself. This ARM approach sounds like a workaround but not like a solution.

**Comment by yochay** [comment]
@mikezks in theory, you are right. However,it is not really practical to expect anything to run for 5 days in the cloud without any 'issues'. Where issue can be VM recycle, any underline dependency of the host recycles, causes the host to hiccup and/ or your code. Regardless, even eventhing works prefect for 2.5 days, and then your code fails, you need to run again for 5 days? If you can break you code apart to where state is 'maintained' so you can handle failures, then you should break your code to run smaller chunks and reduce the 5 days number to minutes or worst hours.

Per @lindydonna , we are looking into this and will update the thread shortly.

**Comment by vhe1** [comment]
My scenario for a longer running function:
**Stream upload**
So, customers upload 50MB and the user experience is to offer this as a continuous stream upload, with the chunking (1MB chunks, then stored in block blobs) done in azure by a function.
Such an upload my take 20min or more.

**Comment by mikezks** [comment]
@yochay: I agree with you that the program code needs to handle unexpected termination of the Function App, but nevertheless it should be possible to define a Function to run longer than 5 min. Then it is up to the developer to repeat certain steps after a termination.

Of course several long running Functions may be avoided by using cloud patters.

I think it is a good idea to collect the different reasons for long running Functions on SO and here in this issue to evaluate whether there is justifiable need for it or not.

**Comment by HWouters** [comment]
I have the following scenario: A few times per year, I have to do many (1000+) computations. Each computation takes a few minutes. To me Functions seemded the perfect solution for this problem, Unfortunately, it didn't work out as expected, because of this maximum time limit of 5 min. Most of the time a single task finishes within the limit, but as soon as I run multiple tasks (I'm using queue trigger), it gets slower and everything times out.

**Comment by paulbatum** [comment]
@HWouters Try using the queues.batchSize setting to reduce how many queue messages a given instance of your function app will process concurrently. You can set the setting via host.json, see here for info:
https://github.com/Azure/azure-webjobs-sdk-script/wiki/host.json

**Comment by dendle** [comment]
I have a scenario which I thought functions (Consumption Plan) was perfect for.

We generate zip files of data to send to third parties every evening via FTP.

I wanted to use a storage queue to instruct an azure function to upload a specified zip file to a specified FTP host.  The 5 minute rule stops this, as FTP upload could take longer.  It's not really suitable to split this job into smaller chunks.

I am currently using cloud service worker roles at the end of a service bus to perform this action.  However, I am struggling not to saturate the (invisible) network bandwidth cap applied to worker role VMs.  I need to send these files as soon as possible, but I cannot accurately scale VMs according to an unkown bandwidth limit on the azure side, and a variable bandwidth of the third party FTP host.

With functions, I wouldn't care.  I'd wire it up to the bus, and the Consumption plan would sort out the compute. I'd care that the files I wanted to FTP were sent.

Perhaps let the user decide the timeout, regardless of maximum length of time.

Thanks for reading,
Matt

**Comment by rhwilburn** [comment]
I really just need 10-15 mins so that I can download a zip file from an FTP and do about 3 mins of processing on it. Currently it takes 3-8 mins to download that file. The FTP server is not my own, so I can't control their slow speeds. The file is ~30mb.

**Comment by ericleigh007** [comment]
I'm running a consumption function (loved the idea of not having to pay for a separate app service plan, would love if everything could eventually be consumption, or at least have that option).  

The function is a on-prem to Office 365 data mover.  it moves the contents of controlled queries from an Azure SQL database to sharepoint lists, used for various purposes.  

Using the 5 minutes limit, the function either has 5 minutes to do all of the table moves, OR I could break it up and allow for a table move per execution, possibly controlled by HTTPs query parameters, but I would still be limited to 5 minutes.

The creation of rows in SharePoint Online is extremely slow, so even lists that are of the "maximum" 5000 rows could cause problems if the maximum runtime is 5 minutes.

I thank you guys for your excellent work, and hope that there will be a happy medium very soon.

**Comment by ericleigh007** [comment]
https://github.com/Azure/azure-webjobs-sdk-script/issues/18 seems to mention are very good idea, although probably not as easy to implement.   That of letting the function request more time.

This gets rid of the problem with arbitrary timeouts by putting the user and programming in charge, and lets the timeout occur by default if nothing happens.

**Comment by cgravill** [comment]
We use Azure Functions with some scientific analysis code, where the majority of the execution time is spent inside a native library (https://github.com/Z3Prover/z3). It would be useful to support running on a consumption with >5 minutes termination even if that increases the probability of something going wrong.

While we may be able to checkpoint progress for possible resumption that might generate a large amount of state in our case.

It’s possible we should actually be using something like Azure Batch or WebJobs for these longer running tasks but the Azure Functions infrastructure is really great to work with. We actually have some tooling around deploying many different versions of our analysis. Each version may only be called very rarely and none might be called on a given day making the consumption model very appealing.

**Comment by donquijotedelamancha** [comment]
I also would need Azure Functions under Consumption plan to run longer than 5 minutes to be able to complete calculation intensive data analytics. 
In my case, switching to webjobs is not an alternative because Azure Functions allow to parallelize runs of the same function whereas this parallelization doesn't seem possible with webjobs.

**Comment by MatthewSteeples** [comment]
This isn't meant to come across as preachy, but Azure Batch sounds like a better fit for a lot of these cases. For some reason it doesn't get all the attention that Functions and WebJobs receive...! We've just finished evaluating the options and differences between the two.  In summary, it supports running multiple versions of an app, automatically scaling up and down (including down to 0 where it won't cost anything) based on a formula, programmatical deployments (we've got ours set up to deploy from VSTS Release Manager), and a greater variety of instance sizes than functions supports. You can also configure it to run more than one task on the same VM (eg in our case we have long running but very lightweight tasks, so we can run 4 on a small VM, rather than 1 per function). Scaling up is limited only by your credit card limit! It's less of a "black box" than functions is, while at the same time not requiring you to manage anything beyond the initial configuration.

The only downside we've found (and it will be acceptable in some cases) is that it can take up to 10 minutes for a new VM to come online. The triggering mechanism is also not as extensive as functions (although you could combine the two and trigger the webjob from a function).

If anyone wants to drop me a message outside of this thread then I'll happily answer questions. Reading through the comments, this seems like it would solve the problems for @HWouters, @dendle, @rhwilburn, @ericleigh007 and @cgravill

**Comment by ericleigh007** [comment]
I think the fact that batch doesn't get the attention is because of both marketing and the documentation.  The documentation for functions makes them look "lightweight" (easy to implement and use, for more simplified capabilities), while the docs for batch seems as if the jobs should be multi-threaded, cored, etc.  The documentation says something like "CPU intensive" tasks.

Toward batch being selected more often for these types of jobs, I think the docs need a few tweaks, so that users would be apt to select them.  The fact that batch jobs seem to be buildable only with Visual Studio also seems to create a higher barrier to entry, compared to the easy web interface for functions.

The comment above about taking 10 minutes to come online means that if a batch wants to react quickly, like a function, it cannot do so.

I realize that Lambda, which is likely the reason that functions was brought out (at least at this point if not at all), only supports 5 minutes as well, but don't we think that a middle of the road between 5 minutes and something else could be just as easily supported on the consumption model?

The worst problem with the 5 minute limitation is that code that is doing maintenance doesn't have ultimate control of how long it will take to do so, and there isn't a way to get extra time, AND there doesn't seem to be a way to allow a function to "continue" itself, in case it does run out of a time.

I'm using an App Service for my function, and that works around the 5 minute limitation, but I'd really like to be consumption based, up to, say... 20 minutes.... an hour.

-thanks
-e

**Comment by HWouters** [comment]
I think we could have used Azure Batch in our project, but it my opinion it isn't as nice and easy to use as functions. For instance it doesn't have the automatic triggering etc. In the project I mentioned before we moved from functions to webjobs, because of this 5 minutes limitation. Too many tasks were timing out, even if batchsize is set to 1. It's unfortunate, because I love the consumption plan from Azure Functions. With webjobs we have to manually scale up/out all the time. 
Does anyone know if this is a temporary limitation which gets resolved in the future? I don't see why you would keep this maximum timeout. I think it's a good default, but why would you prevent users from making it longer?

**Comment by davidtme** [comment]
I’m planning a supervisor / worker etl scenario.  The supervisor will responded to triggers & queues and store it’s state in azure blob/table(?) but my workers can run for an unknown amount of time depending on the work requested so the 5 minute cut off may stop me in my tracks.

**Comment by DarkNoir** [comment]
I think there may be some confusion. There is not a 'time limit' as described. There is a timeout if you attempt a function call that remains open for more than 5 min. **If you make your function async, it will not timeout at 5 min.**

**Comment by DarkNoir** [comment]
I am not hitting this time out in my implementation. I created my function to run a task that takes longer than 5 mins buy using a Task for my logic. The function returns right away obviously since the logic is now async. My async task runs to completion outside the 5 min timeout. Try this and see for yourself.

`DateTime d = DateTime.Now;
Task t = new Task(() =>
{
log.Info("Doing long task");
for (int cnt = 0; cnt < 100; cnt++)
{
log.Info(DateTime.Now.Subtract(d).ToString());
System.Threading.Thread.Sleep(10000);
}
});
t.Start();'

**I am currently running them on a consumption plan for more than 5 mins, as per my post.** There is a timeout, but it’s not on execution time obviously since I am running them longer than 5 mins. The time out is in the calling mechanism that executes the function. If you make it async, the caller returns and the function continues to execute. It is a common practice to put long running tasks on a separate thread/process than the calling thread/process. Keeping the orchestration layer of the function platform from blocking for long periods of time just makes sense. You should not block the calling layer for long running functions as it will seriously degrade the platform.

I wasn't trying to bypass the 5 min timeout, but rather trying to develop with good disciplines. I wanted an async function. I learned about the 'timeout' later. No one should be running long running functions synchronously anyway. Also, people should be aware of how to make async functions regardless of timeouts.

The interwebs can really be a silly place sometimes. I think people are getting so used to cut and pasting code, they are forgetting the basics.

PS. Please add your constraints to this Azure doc:
https://docs.microsoft.com/en-us/azure/azure-subscription-service-limits#app-service-limits

William Wallace
(This is a repost of my post in another location)

**Comment by lindydonna** [comment]
Note that @DarkNoir's approach only works if the Functions host is kept alive through other means. See reply from @mathewc on this thread: https://github.com/Azure/azure-webjobs-sdk-script/issues/18#issuecomment-284049611

**Comment by ericleigh007** [comment]
okay @DarkNoir , please remove your post on [StackOverflow](http://stackoverflow.com/questions/40800933/azure-functions-pricing-and-timeout).   Per what you found above, it is misleading.  I'm glad I came here before bothering with the Task  code.

At any rate, I am still in favor of making Azure "better" than AWS and not just copying their "EXACT" Lambda limitations.

**Comment by DarkNoir** [comment]
Thanks @ericleigh007 for the heads up on being able to delete a post. The post that was deleted was me believing that creating an async function would allow it to run for more than 5 mins. The framework has a timeout internally for the mechanism that calls your functions set at 5 mins. This get bypassed when you make your function run async. HOWEVER there is also a 5 min time limit on how long your App Service instance will be active if there are no OTHER connections keeping it alive. In my case, this second time limit was kept alive by the Azure Portal being connected to it as I was using it for tracing and monitoring the work I was doing. I was getting execution times of almost 2 hours. 

What I learned was AZURE FUNCTIONS DO NOT SUPPORT ASYNC EXECUTION OF THE FUNCTION. They DO SUPPORT MULTITHREADED/ASYNC ACTIVITIES WITHIN YOUR FUNCTION, but, you must synchronize with the main thread before returning to guarantee your function  completes. If you do not, the App Service/VM can be shut down by the orchestration layer since it believes that processing is done since the call returned. 

I really appreciate all the help the community gave me in navigating these elements. I didn't see any previous documents or comments on anyone even attempting async executions. And because of how I was working, I was getting some very confusing and misleading interactions.

The physical repercussions of not being able to call a function async is keeping the connection open for the duration of the call. If you have a long running function and you close the connection before its done, as a minimum, there will be a connection error logged somewhere. I have not researched yet if the function will or could get shut down prematurely also. This means that potentially, you must keep a connection open until the function completes. This can be very bad and dramatically affect the scalability of the App Service that is hosting the function not to mention the application consuming the function.

**Comment by pugwonk** [comment]
As @brettsam and @lindydonna were asking on the older issue (https://github.com/Azure/azure-webjobs-sdk-script/issues/18) for real-life situations that were falling over on the timeout, here's mine.

We've built an Azure function that creates a spreadsheet from JSON data for end-user download, based on a user request from our website. It works great, and functions seemed like a perfect model for us - it may run once in a week, or it may run twenty times in an hour, it depends what our users are doing. For small spreadsheets it's quick, but for large ones it can take ten minutes, quite a lot of which is just upload/download. I see entirely why there's a default limit but I do not see (aside from the VM architecture issues mentioned in the original bug) why I can't increase that. Functions are a great way to run these unpredictably common large jobs and it seems like very much what they're designed for.

I worked for MSFT for ten years and live in Seattle; would be more than happy to pop over and explain my scenario in person if that would help!

**Comment by mathewc** [comment]
This issue has been addressed. See https://github.com/Azure/azure-webjobs-sdk-script/pull/1502

**Comment by pugwonk** [comment]
This is great (and will actually solve my problem) but I think the core issue still remains - there's a hard and somewhat arbitrary limit on how long your function can take, which does rather negate the really appealing "computing time only when you need it" aspect of the Consumption plan. Nobody wants to re-architect their service just because it takes >n minutes on one data set, and changing 5 to 20 doesn't fix that.

I know you have architectural issues making this difficult, but really I should be able to write an Azure function that takes three weeks if I want to pay for that compute time. I'm all in favour of a default limit, but it would be great to see a "real" fix with user-specified limits.

**Comment by BowserKingKoopa** [comment]
I agree.  I have some report generation stuff that can take longer than 20 minutes. WebJobs has no problem handling it.  Why the arbitrary limit in Functions?

**Comment by lindydonna** [comment]
Because of the way the Consumption plan is implemented, we need to be able to move Function Apps from one machine to another. So, like all serverless platforms, we have an execution timeout.

If you want to run for longer, you don't have to re-architect anything, simply run on an App Service Plan instead of the Consumption plan.

**Comment by HWouters** [comment]
@lindydonna, @mathewc Functions are still timing out after 5 minutes: 

_Timeout value of 00:05:00 exceeded by function 'Functions.TriggerRequest' (Id: '1874cd2c-533c-4ccb-9f2c-2434fdfc4bf0'). Initiating cancellation._

**Comment by lindydonna** [comment]
@HWouters The change has not been deployed yet. Watch for a new release on this page: https://github.com/Azure/app-service-announcements/issues

**Comment by hweaver225** [comment]
@lindydonna - We are on an App Service Plan and about 10 days ago started seeing a 5 minute timeout when we had not been experiencing any such timeouts prior.  We are being told we need to rearchitect and stop using Azure functions.  Do you have any insight on this since you indicated we shouldn't see a timeout in an App Service Plan?

**Comment by ericleigh007** [comment]
Make sure  ‘AlwaysOn’ is set in App Service settings.

If you want to use consumption, you currently need to refactor to get smaller bites less than 5  minutes each, just like AWS Lambda.

I have heard nothing from the team about increasing the time limit.

Sent from my Windows 10 phone

**Comment by lindydonna** [comment]
@hweaver225 If you're using an App Service Plan, you can run as long as you want, just make sure "Always On" is on. 

On a Consumption plan, you can run for up to 10 minutes if you set a value for `functionTimeout` in host.json (see https://github.com/Azure/azure-webjobs-sdk-script/wiki/host.json).

The maximum timeout is documented here: https://docs.microsoft.com/en-us/azure/azure-functions/functions-scale#consumption-plan

**Comment by ericleigh007** [comment]
Okay, I had mentioned a need for longer running functions, but since that I did a complete refactor of my workloads and came up with this.
Timer function sets the whole thing in motion
- Looks at a master job list (in this case a SharePoint List), which can have snippets of stuff that need to be done
- For each entry in the list that is enabled, queues out an entry on an Azure Queue, in this case using a serialized json dictionary (but of course could be a custom object)

Queue function for each piece of work
- Sees a new queue element and decodes the element to see what work is needed.
- The queue elements carry with them a startElement, for restartability
- In my case, the operation usually requires deleting some previous output, which is long running.  I do this work in chunks of a size that I defined using Environment variables
- Then the job copies in new input to the output.
- During the execution of anything long-running, I have a little check function that is run when small chunks of processing are done (think 100 or 200 rows of copying into a sharepoint list, for instance) that sees if I'm out of time, based upon another Environment Variable that tells me how much time to allot.
- If I'm out of time, the function reports that and the callers is notified to chain to a continuation job
- The continuation, job is queued, and the next "chain element" takes over.

This way I never run out of time.  Queue elements keep the thing very modular and efficient.  And using queue elements means that if something crashes, it gets retried without writing any code.  It also gets put on a poison queue if it fails completely, and one can easily use that queue to blast out emails or SMS or push notifications without code using Logic Apps.

One particular thing to note (which is sort of contrary to 12 factor Apps ) is that functions do run in multiple threads and NOT processes.  This means if you're in C#, using` ``[ThreadStatic]`` `for anything that needs to be thread safe, there are many ways based on the language you're using.  

In short, now that I took the time to redo everything, I'm very happy with functions, and can run completely on consumption (which saved mucho $$$$).

Enjoy!

**Comment by mikezks** [comment]
@ericleigh007: have you already tried to use the new [Durable Functions Framework](https://blogs.msdn.microsoft.com/appserviceteam/2017/07/06/alpha-preview-for-durable-functions/)? Maybe this helps too.
